File: csharp/word-count/WordCount.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace phrase
{
	public class Phrase
	{
		public static IDictionary<string, int> WordCount (string phrase)
		{
			var counts = new Dictionary<string, int> ();
			Match match = FindMatch(phrase);
			while (match.Success) {
				string word = match.Value;
				if (!counts.ContainsKey (word))
				{
					counts [word] = 0;
				}
				counts [word]++;
				match = match.NextMatch ();
			}
			return counts;
		}

		private static Match FindMatch (string phrase)
		{
			return Regex.Match (phrase.ToLower (), @"\w+'\w+|\w+");
		}
	}
}

File: ruby/word-count/word_count.rb
module BookKeeping
  VERSION = 1
end

class Phrase
  def initialize(phrase)
    @phrase = phrase
  end

  def word_count
    counts = Hash.new(0)
    words.each do |word|
      counts[word] += 1
    end
    counts
  end

  private

  attr_reader :phrase

  def words
    phrase.downcase.scan(/\b[\w+']+\b/)
  end
end